server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesContext.java

```diff
@@ -33,20 +32,25 @@
 
     private final List<FieldAndFormat> fields;
 
-    public static FetchDocValuesContext create(Function<String, Set<String>> simpleMatchToFullName,
-                                               int maxAllowedDocvalueFields,
-                                               List<FieldAndFormat> fieldPatterns) {
+    public static FetchDocValuesContext create(MapperService mapperService, List<FieldAndFormat> fieldPatterns) {
+        int maxAllowedDocvalueFields = mapperService.getIndexSettings().getMaxDocvalueFields();
+
         List<FieldAndFormat> fields = new ArrayList<>();
+        int mappedFieldCount = 0;
         for (FieldAndFormat field : fieldPatterns) {
-            Collection<String> fieldNames = simpleMatchToFullName.apply(field.field);
+            Collection<String> fieldNames = mapperService.simpleMatchToFullName(field.field);
             for (String fieldName: fieldNames) {
                 fields.add(new FieldAndFormat(fieldName, field.format));
+                // only count the mapped fields towards the limit fedinde by IndexSettings.MAX_DOCVALUE_FIELDS_SEARCH_SETTING
+                if (mapperService.fieldType(fieldName) != null) {
+                    mappedFieldCount++;
+                }
             }
         }
-        if (fields.size() > maxAllowedDocvalueFields) {
+        if (mappedFieldCount > maxAllowedDocvalueFields) {
             throw new IllegalArgumentException(
-                "Trying to retrieve too many docvalue_fields. Must be less than or equal to: [" + maxAllowedDocvalueFields
-                    + "] but was [" + fields.size() + "]. This limit can be set by changing the ["
+                "Trying to retrieve too many mapped docvalue_fields. Must be less than or equal to: [" + maxAllowedDocvalueFields
```

Comment:
For me the old message was just as clear, and lets us avoid having to update the REST tests.